sdks/java/io/hadoop-common/src/main/java/org/apache/beam/sdk/io/hadoop/SerializableConfiguration.java

```diff
@@ -49,34 +52,40 @@ public SerializableConfiguration(Configuration conf) {
     if (conf == null) {
       throw new NullPointerException("Configuration must not be null.");
     }
+    this.confMutated = true;
     this.conf = conf;
   }
 
   public Configuration get() {
-    if (serializationCache != null) {
-      serializationCache = null;
-    }
+    // get() call returns the original conf, which is mutable by caller
+    confMutated = true;
     return conf;
   }
 
   @Override
   public void writeExternal(ObjectOutput out) throws IOException {
-    if (serializationCache == null) {
-      ByteArrayOutputStream baos = new ByteArrayOutputStream(512);
-      try (DataOutputStream dos = new DataOutputStream(baos)) {
-        conf.write(dos);
-        serializationCache = baos.toByteArray();
+    String canonicalName;
+    synchronized (this) {
+      canonicalName = conf.getClass().getCanonicalName();
+      if (confMutated) {
+        confMutated = false;
+        ByteArrayOutputStream baos = new ByteArrayOutputStream(512);
+        try (DataOutputStream dos = new DataOutputStream(baos)) {
+          // this call is slow.
+          conf.write(dos);
+          serializationCache = baos.toByteArray();
```

Comment:
This is the only line that needs to be synchronized I think

Reply:
Actually with this implementation, synchronization is not required unless `conf.write` will explode if there is some concurrent modification.

Reply:
I was thinking about the possible racing condition that writeExternal could write name and bytes from different conf, and ended up including much of the weiteExternal in synchronized block.

Added more comments about the intentions.

Reply:
> Actually with this implementation, synchronization is not required unless conf.write will explode if there is some concurrent modification.

Yeah thinking again if conf is modified half way the whole writeExternal won't work anyways, going to remove the synchronized block